src/Test.EndToEnd.Baselining/Extensions/ArtifactUriEnumeration.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Sarif;
+
+namespace SarifBaseline.Extensions
+{
+    public static class ArtifactUriEnumeration
+    {
+        public static IEnumerable<Uri> AllArtifactUris(this SarifLog log)
+        {
+            if (log != null)
+            {
+                foreach (Run run in log.EnumerateRuns())
+                {
+                    foreach (Uri uri in run.AllArtifactUris())
+                    {
+                        yield return uri;
+                    }
+                }
+            }
+        }
+
+        public static IEnumerable<Uri> AllArtifactUris(this Run run)
+        {
+            if (run?.Artifacts != null)
+            {
+                foreach (Artifact artifact in run.Artifacts)
+                {
+                    Uri uri = artifact?.Location?.Uri;
+                    if (uri != null) { yield return uri; }
+                }
+            }
+
+            foreach (Result result in run.EnumerateResults())
+            {
+                foreach (Uri uri in AllArtifactUris(result))
+                {
+                    yield return uri;
+                }
+            }
+        }
```

Comment:
>} [](start = 8, length = 1)

The problem with this approach to enumeration is that `artifactLocation` objects can occur in many `run` properties other than `result` -- and they can occur in several `result` properties as well. Below you will find a list -- which I don't guarantee to be exhaustive! -- based on searching the spec for the string "artifactLocation object".

IMO you are better off creating a visitor based on `SarifRewritingVisitor`, something like `ArtifactLocationFindingVisitor`, that accumulates all the `artifactLocation` objects it encounters, and then at the end you can run over them:

```
var visitor = new ArtifactLocationFindingVisitor();
visitor.Visit(sarifLog) // or run, or whatever you're interested in.
List<Uri> artifactUris = visitor.ArtifactLocations
    .Where(a => a.Uri != null)
    .Select(a => a.Uri)
    .Distinct()
    .ToList();
```

All of these property paths are relative to a `run` object

- `attachments[].location`
- `conversion.analysisToolLogFiles`
- `conversion.tool.driver.locations`
- `conversion.tool.extensions[].locations`
- `externalPropertyFileReferences[].location`
- `invocations[].responseFiles`
- `invocations[].executableLocation`
- `invocations[].workingDirectory`
- `invocations[].stdin, stdout, stderr, stdoutStderr`
- `originalUriBaseIds[]`
- `policies[].locations`
- `specialLocations[].displayBase`
- `stacks[].frames[].location.physicalLocation.artifactLocation`
- `taxonomies[].locations`
- `threadFlowLocations[].location.physicalLocation.artifactLocation`
- `tool.driver.locations`
- `tool.extensions[].locations`
- `translations[].locations`
- `result.attachments[].location`
- `versionControlProvenance[].mappedTo`

Reply:
Of course if the semantics you want is "all result location URIs" then this is fine -- but then you need to name the methods to make clear what they're providing. It's not "all artifact URIs in the result".

---
In reply to: [366541420](https://github.com/microsoft/sarif-sdk/pull/1734#discussion_r366541420) [](ancestors = 366541420)

Reply:
Larry's suggestion, of course, will collect all these things in memory at once, which might not be ideal. I like the visitor pattern but am not sure it's the right approach (that would depend on consumer use case). I do agree with Larry's core point that, you need to cover other url holding things. It does raise the point, why was this code authored to begin with? i.e., what was the consumer doing with the information? are consumers likely to only want the unique urls or all of them? also, do they want reconstructed URLs if there is a original uri base?


---
In reply to: [366542388](https://github.com/microsoft/sarif-sdk/pull/1734#discussion_r366542388) [](ancestors = 366542388,366541420)

Reply:
Hmm. My usage needs all the Locations in all the Results, so I'll go with the rename the method option. =)

Reply:
I won't move these to the SDK. 

Further, I've renamed the ones above Result scope to be "AllResultArtifactUris" and made the Result one include Result.Attachments[].Location, to be complete.

I think technically my scenario doesn't care if it's everything in the Run or just the Result.Locations Uris, so once a variation of this in the SDK is added, I should be able to use it.

Larry, I like the Visitor accumulator idea; that seems like a nice, consistent way to get them all. I'm going to leave my code as-is since it's done and won't be ported to a broader location, however.